File: android/app/src/main/java/com/figural/app/manager/GlassesManager.kt
package com.figural.app.manager

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.meta.wearable.dat.core.Device
import com.meta.wearable.dat.core.PermissionStatus
import com.meta.wearable.dat.core.PermissionType
import com.meta.wearable.dat.core.RegistrationState
import com.meta.wearable.dat.core.Wearables
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class GlassesManager(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val wearables = Wearables.getInstance()

    private val _isRegistered = MutableStateFlow(false)
    val isRegistered: StateFlow<Boolean> = _isRegistered.asStateFlow()

    private val _isRegistering = MutableStateFlow(false)
    val isRegistering: StateFlow<Boolean> = _isRegistering.asStateFlow()

    private val _devices = MutableStateFlow<List<Device>>(emptyList())
    val devices: StateFlow<List<Device>> = _devices.asStateFlow()

    private val _cameraPermissionStatus = MutableStateFlow(PermissionStatus.DENIED)
    val cameraPermissionStatus: StateFlow<PermissionStatus> = _cameraPermissionStatus.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    val hasCameraPermission: Boolean
        get() = _cameraPermissionStatus.value == PermissionStatus.GRANTED

    init {
        observeRegistrationState()
        observeDevices()
    }

    private fun observeRegistrationState() {
        scope.launch {
            wearables.registrationStateFlow().collect { state ->
                when (state) {
                    RegistrationState.REGISTERED -> {
                        _isRegistered.value = true
                        _isRegistering.value = false
                        checkCameraPermission()
                    }
                    RegistrationState.UNREGISTERED -> {
                        _isRegistered.value = false
                        _isRegistering.value = false
                    }
                    RegistrationState.REGISTERING -> {
                        _isRegistering.value = true
                    }
                    RegistrationState.UNREGISTERING -> {
                        _isRegistering.value = true
                    }
                }
            }
        }
    }

    private fun observeDevices() {
        scope.launch {
            wearables.devicesFlow().collect { deviceList ->
                _devices.value = deviceList
            }
        }
    }

    fun register() {
        try {
            _isRegistering.value = true
            _error.value = null
            wearables.startRegistration(context)
        } catch (e: Exception) {
            _isRegistering.value = false
            _error.value = e.message ?: "Registration failed"
        }
    }

    fun unregister() {
        try {
            wearables.startUnregistration(context)
        } catch (e: Exception) {
            _error.value = e.message ?: "Unregistration failed"
        }
    }

    suspend fun handleCallback(uri: Uri) {
        try {
            wearables.handleUrl(uri)
        } catch (e: Exception) {
            _error.value = e.message ?: "Failed to handle callback"
        }
    }

    suspend fun checkCameraPermission() {
        try {
            _cameraPermissionStatus.value = wearables.checkPermissionStatus(PermissionType.CAMERA)
        } catch (e: Exception) {
            _cameraPermissionStatus.value = PermissionStatus.DENIED
        }
    }

    suspend fun requestCameraPermission() {
        try {
            _cameraPermissionStatus.value = wearables.requestPermission(context, PermissionType.CAMERA)
        } catch (e: Exception) {
            _cameraPermissionStatus.value = PermissionStatus.DENIED
            _error.value = e.message ?: "Failed to request camera permission"
        }
    }

    fun clearError() {
        _error.value = null
    }
}

File: Figural/Managers/GlassesManager.swift
import Foundation
import SwiftUI
import MWDATCore

@MainActor
final class GlassesManager: ObservableObject {
    @Published var isRegistered: Bool = false
    @Published var devices: [Device] = []
    @Published var cameraPermissionStatus: PermissionStatus = .denied
    @Published var registrationError: Error?
    @Published var isRegistering: Bool = false
    
    private let wearables = Wearables.shared
    
    init() {
        Task {
            await configure()
            await observeRegistrationState()
            await observeDevices()
        }
    }
    
    private func configure() async {
        do {
            try Wearables.configure()
        } catch {
            print("Failed to configure Wearables: \(error)")
        }
    }
    
    private func observeRegistrationState() async {
        for await state in wearables.registrationStateStream() {
            switch state {
            case .registered:
                self.isRegistered = true
                self.isRegistering = false
                await checkCameraPermission()
            case .unregistered:
                self.isRegistered = false
                self.isRegistering = false
            case .registering:
                self.isRegistering = true
            case .unregistering:
                self.isRegistering = true
            @unknown default:
                break
            }
        }
    }
    
    private func observeDevices() async {
        for await deviceList in wearables.devicesStream() {
            self.devices = deviceList
        }
    }
    
    func register() {
        do {
            isRegistering = true
            registrationError = nil
            try wearables.startRegistration()
        } catch {
            isRegistering = false
            registrationError = error
            print("Registration failed: \(error)")
        }
    }
    
    func unregister() {
        do {
            try wearables.startUnregistration()
        } catch {
            print("Unregistration failed: \(error)")
        }
    }
    
    func handleCallback(url: URL) async {
        do {
            _ = try await wearables.handleUrl(url)
        } catch {
            print("Failed to handle callback URL: \(error)")
            registrationError = error
        }
    }
    
    func checkCameraPermission() async {
        do {
            cameraPermissionStatus = try await wearables.checkPermissionStatus(.camera)
        } catch {
            print("Failed to check camera permission: \(error)")
            cameraPermissionStatus = .denied
        }
    }
    
    func requestCameraPermission() async {
        do {
            cameraPermissionStatus = try await wearables.requestPermission(.camera)
        } catch {
            print("Failed to request camera permission: \(error)")
            cameraPermissionStatus = .denied
        }
    }
    
    var hasCameraPermission: Bool {
        cameraPermissionStatus == .granted
    }
}
